### DISCUSSION TITLE
Stream and Save

### DISCUSSION THREAD

**Discussion by galileor97** [iina/iina#4859] [General]
Can I use iina to stream and save videos? I can't seem to find the save feature in the settings.

**Comment by low-batt** [comment]
For audio/video support IINA uses a library from the [mpv project](https://github.com/mpv-player/mpv). It is `mpv` that obtains bytes from [yt-dlp](https://github.com/yt-dlp/yt-dlp) and forms the video frames that IINA then displays.

So the question is whether `mpv` can even do this. I searched the net and people recommended using the `mpv` [stream-record](https://mpv.io/manual/stable/#options-stream-record) option. <strike>_However_ my expectation is that this option is unlikely to work with IINA. IINA is a media player and is not intended to encode and write video files. The [FFmpeg](https://ffmpeg.org/) libraries contained in IINA have not been built with support for encoding enabled.</strike> **UPDATE** @mcstrassell09 shows how to accomplish this in the comments below.

Many thanks to @mcstrassell09 for the instructions on how to use `stream-record`.

If I want to retain a video I first download it using [yt-dlp](https://github.com/yt-dlp/yt-dlp) and then play the video from the local copy on disk. This also has the advantage of confirming the local copy is indeed playable and complete.

**Comment by mcstrassell09** [comment]
Yes, you can  record and save streaming videos with IINA by using the mpv stream-record option with a custom lua script and key binding.  Here is a lua script called streamRecord.lua. Place this in your ~/.config/mpv/scripts folder. Then you must add the necessary key-bindings in IINA settings: 


<img width="681" alt="Screenshot 2024-12-16 at 1 58 01 AM" src="https://github.com/user-attachments/assets/53040a61-666c-461d-974f-b31f565cfd40" />

For this to work, IINA must be opened from a writeable directory. it must be opened from terminal or iterm after cd'ing to a writeable directory.  For example:

```
cd ~/videos
/Applications/IINA.app/Contents/MacOS/iina-cli --mpv-mute=yes --mpv-profile=enc-v-h264 ~/playlists/chat.m3u8
```

Creating m3u8 playlists is another story. Do a little research. Here is streamRecord.lua

```
local mp = require 'mp'
local utils = require 'mp.utils'
local msg = require 'mp.msg'
local assdraw = require 'mp.assdraw'

function stream_recorder()
	local playlist_pos = mp.get_property("playlist-playing-pos")
	local playlist_title_string = "playlist/"..playlist_pos.."/title"
	local playlist_title = mp.get_property(playlist_title_string)
	local media_title = mp.get_property("media-title")
	local clockmins = mp.get_property("clock")
	
	if playlist_title
	then
		
		local filename = playlist_title..clockmins..".mp4"
		mp.set_property("stream-record", filename)
		mp.osd_message("Recording file "..filename)
	else
		local cfilename = media_title..clockmins..".mp4"
		mp.set_property("stream-record", cfilename)
		mp.osd_message("Recording file "..cfilename)
	end
end

function stop_recorder()
	mp.set_property("stream-record", "")
	mp.osd_message("Recording stopped")
end

mp.add_key_binding("r", "stream-recorder", stream_recorder)
mp.add_key_binding("R", "stop-recorder", stop_recorder)
```

This is a hack to make the mpv stream-record option work with lua scripting in IINA. I am not a trained programmer. Use at your own risk. Also, hire me. Thanks.

**Reply by mcstrassell09** [reply]
Here's a hack to create an m3u8 live playlist with yt-dlp that then can be opened with IINA. 
 
```
#!/bin/bash
min=1
max=3
header="#EXTM3U"
urlfile=$HOME/video-processing/playlists/chat.m3u8
hlsfile=$HOME/video-processing/playlists/Untitled.m3u
intfile=$HOME/video-processing/playlists/int.m3u
tmpfile=$HOME/video-processing/playlists/tmp.json
url="$(pbpaste)"
title="$(echo "$url" | cut -d'/' -f4)"
ytdlcmd="$(yt-dlp --no-warnings -J --flat-playlist --no-playlist -- $url | tee $tmpfile)"

#$ytdlcmd

if [ -f "$hlsfile" ]; then
	echo "$header" > "$intfile"
	echo "$url" >> "$urlfile"
	for i in $(seq $min $max)
	do
		echo "#EXTINF:-1,"$title"_""$i" | tee -a "$hlsfile" >> "$intfile"
		cat $tmpfile | jq -r '.formats' | jq ".[$i].url" | tr -d \" | tee -a $hlsfile >> $intfile
	done
	echo "$title"
else
	echo "$header" | tee "$hlsfile" "$intfile"
	echo "$url" >> "$urlfile"
	for i in $(seq $min $max)
	do
		echo "#EXTINF:-1,"$title"_""$i" | tee -a $hlsfile >> $intfile
		cat $tmpfile | jq -r '.formats' | jq ".[$i].url" | tr -d \" | tee -a $hlsfile >> $intfile
	done
	echo "$title"
fi


```

add this to your macOS Shortcuts App (adjust directory as needed). After you cmd-c or copy the intended link, the shortcut  will run on success. You may need to install JQ with homebrew: "brew install jq". Once IINA is playing the llve stream, pressing key binding "r" will activate mpv stream-record option and save to current writeable working-directory. Pressing keybinding "R" that is "Shift + r" will stop stream-record and complete the stream-save.

**Comment by mcstrassell09** [comment]
You can also add an mpv watch property to see what worrking-directory IINA was launched from and if mpv stream-record is active from the IINA inspector window

<img width="996" alt="Screenshot 2024-12-16 at 2 56 11 AM" src="https://github.com/user-attachments/assets/2bd990d7-823b-4901-a103-c665ba51ae37" />

Here you can see IINA was launched from a writeable working-directory and the mpv stream-record option is active.
